manager/controllers/app/select_modules.go

```diff
@@ -116,51 +115,81 @@ func (p *PlotterGenerator) validSolutions(item *DataInfo, solutions []Solution,
+// where storage can be allocated
+// what additional actions to perform
+func (p *PlotterGenerator) validateStorageRequirements(item *DataInfo, application *app.FybrikApplication, element *ResolvedEdge) bool {
 	appContext := ApplicationContext{Application: application, Log: p.Log}
-	requiredActions := item.Actions
-	for _, element := range solution.DataPath {
-		element.Actions = []taxonomy.Action{}
-		moduleCapability := element.Module.Spec.Capabilities[element.CapabilityIndex]
-		if !element.Edge.Sink.Virtual && (item.Context.Flow != taxonomy.WriteFlow ||
-			item.Context.Requirements.FlowParams.IsNewDataSet) {
-			// storage is required, plus more actions on copy may be needed
-			isAccountFound := false
-			for accountInd := range p.StorageAccounts {
-				// validate restrictions
-				if !p.validateRestrictions(item.Configuration.ConfigDecisions[moduleCapability.Capability].
-					DeploymentRestrictions.StorageAccounts, &p.StorageAccounts[accountInd].Spec,
-					p.StorageAccounts[accountInd].Name) {
-					p.Log.Debug().Str(logging.DATASETID, item.Context.DataSetID).Msgf("storage account %s does not match the requirements",
-						p.StorageAccounts[accountInd].Name)
-					continue
-				}
-				// query the policy manager whether WRITE operation is allowed
-				operation := new(policymanager.RequestAction)
-				operation.ActionType = taxonomy.WriteFlow
+	var err error
+	var actions []taxonomy.Action
+	operation := &policymanager.RequestAction{}
+	operation.ActionType = taxonomy.WriteFlow
+
+	if item.Context.Flow == taxonomy.WriteFlow && !item.Context.Requirements.FlowParams.IsNewDataSet {
+		// no need to allocate storage, write destination is known
+		// query the policy manager whether WRITE operation is allowed
+		operation.Destination = item.DataDetails.ResourceMetadata.Geography
+		operation.ProcessingLocation = taxonomy.ProcessingLocation(item.WorkloadCluster.Metadata.Region)
+		actions, err = LookupPolicyDecisions(item.Context.DataSetID, p.PolicyManager, appContext, operation)
+		if err != nil {
+			p.Log.Error().Err(err).Msg("could not validate governance action for writing data")
+			return false
+		}
+	} else {
+		for accountInd := range p.StorageAccounts {
+			// validate restrictions
+			moduleCapability := element.Module.Spec.Capabilities[element.CapabilityIndex]
+			if !p.validateRestrictions(
+				item.Configuration.ConfigDecisions[moduleCapability.Capability].DeploymentRestrictions.StorageAccounts,
+				&p.StorageAccounts[accountInd].Spec,
+				p.StorageAccounts[accountInd].Name) {
+				p.Log.Debug().Str(logging.DATASETID, item.Context.DataSetID).Msgf("storage account %s does not match the requirements",
+					p.StorageAccounts[accountInd].Name)
+				continue
+			}
+			// query the policy manager whether WRITE operation is allowed
+			if !item.Context.Requirements.FlowParams.IsNewDataSet {
 				operation.Destination = string(p.StorageAccounts[accountInd].Spec.Region)
```

Comment:
I'm wondering if we should add a function in the data governance engine connector that does some of this work.  I.e. receives a dataset id, flow, and a list of locations (storage locations) and returns which are allowed.  Wouldn't that be cleaner that doing it here?

Reply:
Yes, it will. The output should include a list of actions per each location, since storing in some may require to reduce a column, storing in another can be done as-is, ...